cache-redis.go

```diff
@@ -50,18 +152,35 @@ func (b *redisBackend) Lookup(q *dns.Msg) (*dns.Msg, bool, bool) {
 	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
 	defer cancel()
 	key := b.keyFromQuery(q)
-	value, err := b.client.Get(ctx, key).Result()
+
+	// Fetch raw bytes to avoid string conversion overhead
+	valueBytes, err := b.client.Get(ctx, key).Bytes()
 	if err != nil {
 		if errors.Is(err, redis.Nil) { // Return a cache-miss if there's no such key
 			return nil, false, false
 		}
 		Log.Error("failed to read from redis", "error", err)
 		return nil, false, false
 	}
+
+	// Try binary decode first, with JSON fallback for backward compatibility
 	var a *cacheAnswer
-	if err := json.Unmarshal([]byte(value), &a); err != nil {
-		Log.Error("failed to unmarshal cache record from redis", "error", err)
-		return nil, false, false
+	a, err = decodeCacheAnswer(valueBytes)
+	if err != nil {
+		// Fallback to JSON for backward compatibility with existing cached entries
+		// Quick heuristic: JSON likely starts with '{'
+		if len(valueBytes) >= 1 && valueBytes[0] == '{' {
+			if jsonErr := json.Unmarshal(valueBytes, &a); jsonErr != nil {
+				Log.Error("failed to decode cache record from redis (binary and JSON)", "binary_error", err, "json_error", jsonErr)
+				return nil, false, false
+			}
+		} else {
+			// Try JSON anyway
+			if jsonErr := json.Unmarshal(valueBytes, &a); jsonErr != nil {
+				Log.Error("failed to decode cache record from redis", "error", err)
+				return nil, false, false
+			}
+		}
```

Comment:
The conditional here looks unnecessary, it'll unmarshal with JSON regardless, if it starts with `{` or not.